File: leetcode_cpp/27_移除元素.cpp
//
// Created by admin on 2024/7/8.
//
#include <vector>
using namespace std;

class Solution {
public:
    // 构建新数组
    int removeElement1(vector<int>& nums, int val) {
        vector<int> new_nums(nums.size(), 0);
        int new_nums_index = 0;
        for (int i = 0; i < nums.size(); ++i) {
            if (nums[i] != val) {
                new_nums[new_nums_index++] = nums[i];
            }
        }
        nums = new_nums;
        return new_nums_index;
    }
    // 双指针法
    int removeElement2(vector<int>& nums, int val) {
        int left = 0, right = 0;
        for (; right < nums.size(); right++) {
            if (val != nums[right]) {
                nums[left++] = nums[right];
            }
        }
        return left;
    }
};

File: leetcode_rust/src/27_移除元素.rs
impl Solution {
    // 建立新数组
    pub fn remove_element1(nums: &mut Vec<i32>, val: i32) -> i32 {
        let mut new_nums: Vec<i32> = vec![0; nums.len()];
        let mut new_nums_index: usize = 0;
        for i in 0..nums.len() {
            if nums[i] != val {
                new_nums[new_nums_index] = nums[i];
                new_nums_index += 1;
            }
        }
        *nums = new_nums;
        return new_nums_index as i32;
    }
    // 双指针法
    pub fn remove_element2(nums: &mut Vec<i32>, val: i32) -> i32 {
        let mut slow_index = 0_usize;

        for fast_index in 0..nums.len() {
            if nums[fast_index] != val {
                nums[slow_index] = nums[fast_index];
                slow_index += 1;
            }
        }
        return slow_index as i32;
    }
}